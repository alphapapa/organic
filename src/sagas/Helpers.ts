// * Helpers.ts<sagas>

// ** License

/**
 * Copyright (C) 2018, Bart≈Çomiej Nankiewicz<bartlomiej.nankiewicz@gmail.com>
 *
 * This file is part of Organic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// ** Tasks

// *** TODO Add types

// * Imports

import R from 'ramda';
import { call } from 'redux-saga/effects';
import DialogAndroid from 'react-native-dialogs';

// * Code

export const prepareNoteForDb = R.evolve({
  drawers: (drawers) =>
    typeof drawers === 'object' ? JSON.stringify(drawers) : drawers,
  tags: R.map((tag) => ({ name: tag, isContextTag: false })),
});

export function* alert(title, msg) {
  yield call(DialogAndroid.alert, title, msg);
}

export function* confirmDialog(title, msg) {
  const { action } = yield call(DialogAndroid.alert, title, msg, {
    positiveText: 'yes',
    negativeText: 'no',
  });

  switch (action) {
    case DialogAndroid.actionPositive:
      return true;
  }
}

export function* textInputDialog(title, msg) {
  const { action, text } = yield call(DialogAndroid.prompt, title, msg, {
    neutralText: 'cancel',
  });

  switch (action) {
    case DialogAndroid.actionPositive:
      return text;
    default:
  }
}
