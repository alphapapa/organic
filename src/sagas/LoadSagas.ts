// * LoadSagas.ts

// ** License

/**
 * Copyright (C) 2018, Bart≈Çomiej Nankiewicz<bartlomiej.nankiewicz@gmail.com>
 *
 * This file is part of Organic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// ** Tasks

// *** TODO Add types

// *** TODO Add docs

// * Imports

import { OrgApi } from 'org-mode-connection';
import R from 'ramda';
import { call, put, select } from 'redux-saga/effects';
import NavigationRedux from '../redux/NavigationRedux';
import OrgDataRedux, { OrgDataSelectors } from '../redux/OrgDataRedux';
import SyncRedux from '../redux/SyncRedux';
import { getFileTitle } from '../utils/files';
import { vibrate } from '../vibrations';
import NodeSagas from './NodeSagas';

// * Code

export function* rehydrateNewNodeCache(node) {
  yield* NodeSagas.updateNode({ node });
  const fileNodes = yield select(OrgDataSelectors.getFileNodes(node.fileId));
  const newIds = R.pipe(
    R.values,
    R.sortBy(R.prop('position')),
    R.map(R.prop('id')),
  )(fileNodes);
  const newTOC = R.pipe(R.filter(R.propEq('level', 1)), R.map(R.prop('id')))(
    fileNodes,
  );
  yield put(OrgDataRedux.updateFileIds(node.fileId, newIds, newTOC));
}

function* getOrCreateNode(target) {
  const [node, created] = yield call(OrgApi.getOrCreateNodeByHeadline, target);
  if (created) {
    yield* rehydrateNewNodeCache(node);
  }
  return node.id;
}

export function* loadAgenda(action) {
  const range = yield select(OrgDataSelectors.getTimestampsRange);
  const { agendaItems, dayAgendaItems, nodes } = yield call(
    OrgApi.getAgendaAsPlainObject,
    range,
  );
  yield* loadNodes(nodes, 'agenda');
  yield put(OrgDataRedux.updateTimestamps(agendaItems, dayAgendaItems, range));
}

export function* loadBreadcrumbs(action) {
  const { target } = action.payload;
  let file;
  let ancestors = [];

  file = yield select(OrgDataSelectors.getFile(target.fileId));

  if (target.nodeId || target.headline) {
    const nodeId = yield* getOrCreateNode(target);
    ancestors = yield call(OrgApi.getAncestorsAsPlainObject, nodeId);
  }

  yield put(
    NavigationRedux.updateNavigation({
      breadcrumbs: [
        {
          fileId: target.fileId,
          headline: getFileTitle(file),
        },
        ...R.map(R.pick(['id', 'headline']), ancestors),
      ],
    }),
  );
}

export function* loadNodes(nodesList, stack) {
  yield put(OrgDataRedux.updateLoadedNodes(R.indexBy(R.prop('id'), nodesList)));

  yield put(OrgDataRedux.openIds(nodesList.map((o) => o.id), stack));
  yield put(OrgDataRedux.stackLoaded(stack));
}

function* loadOrgData(updateChangedFiles = true) {
  const files = yield call(OrgApi.getAllFilesAsPlainObject);
  const tags = yield call(OrgApi.getTagsAsPlainObject);
  const tocs = yield call(OrgApi.getTocs);

  // Load Files and Tags
  yield put(
    OrgDataRedux.getOrgDataSuccess(
      R.map(R.omit(['nodesData', 'isChanged']), files),
      tags,
    ),
  );

  // Rehydrate locally changed files
  if (updateChangedFiles) {
    const filesIds = R.into(
      [],
      R.compose(
        R.filter(R.propEq('isChanged', true)),
        R.map((file) => file.id),
      ),
      files,
    );

    yield put(SyncRedux.markAsChanged({ filesIds }));
  }
  // Load TOCS
  for (var i = 0; i < files.length; i++) {
    const fileId = files[i].id;
    yield put(OrgDataRedux.updateFileIds(fileId, null, tocs.ids[fileId]));
    yield put(OrgDataRedux.updateLoadedNodes(tocs.data));
  }

  // Load Pinned Nodes
  yield* loadPinnedNodes();
  yield put(OrgDataRedux.loadWeekAgenda(new Date()));
}

export function* loadPinnedNodes() {
  const pinned = yield call(OrgApi.search, {
    todos: {},
    tags: { PIN: 1 },
    priority: {},
    isScheduled: false,
    hasDeadline: false,
    searchTerm: '',
  });

  yield put(
    OrgDataRedux.updateLoadedNodes(
      R.indexBy(R.prop('id'), pinned),
      'markedPlaces',
    ),
  );

  yield put(OrgDataRedux.openIds(pinned.map((o) => o.id), 'markedPlaces'));
}

export function* loadRelatedNodes({ nodeId }) {
  const nodesList = yield call(OrgApi.getRelatedNodes, nodeId);

  yield* loadNodes(nodesList);
}

export function* loadTOC({ fileId, navigator }) {
  vibrate();
  yield call(navigator.handleDeepLink, { link: `notes/${fileId}` });
}

// * Export

export default {
  getOrCreateNode,
  loadAgenda,
  loadBreadcrumbs,
  loadNodes,
  loadOrgData,
  loadPinnedNodes,
  loadRelatedNodes,
  loadTOC,
};
