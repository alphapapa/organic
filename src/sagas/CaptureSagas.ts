// * CaptureSagas.ts

// ** License

/**
 * Copyright (C) 2018, Bartłomiej Nankiewicz<bartlomiej.nankiewicz@gmail.com>
 *
 * This file is part of Organic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// ** Tasks

// *** TODO Add types

// * Imports

import { OrgApi } from 'org-mode-connection';
import R from 'ramda';
import { Keyboard } from 'react-native';
import DialogAndroid from 'react-native-dialogs';
import { getFormValues } from 'redux-form';
import { call, put, select } from 'redux-saga/effects';
import { NavigationActions } from '../navigation';
import CaptureActions, { CaptureSelectors } from '../redux/CaptureRedux';
import OrgDataRedux, { OrgDataSelectors } from '../redux/OrgDataRedux';
import { Colors } from '../themes';
import { prepareNoteForDb } from './Helpers';
import NodeSagas from './NodeSagas';

// * Code

function* addCaptureTemplate(raction) {
  const { target, after } = raction.payload;
  let ctTarget;
  let ctContent;
  if (target.target) {
    ctTarget = target.target;
    ctContent = R.omit(['target', 'name'], target);
    ctName = target.name;
  } else {
    // Only target for new template
    ctTarget = target;
    ctContent = {};
  }
  const { action, text } = yield call(
    DialogAndroid.prompt,
    'Add Capture Template',
    'Enter name for new capture template:',
    {
      neutralText: 'Cancel',
    },
  );

  switch (action) {
    case DialogAndroid.actionPositive:
      if (R.isNil(text)) {
        yield call(DialogAndroid.alert, 'Add Capture Template', 'Empty name');
      } else {
        const exists = yield select(
          CaptureSelectors.captureTemplateExists(text),
        );
        if (!exists) {
          yield put(
            CaptureActions.addCaptureTemplate.success({
              name: text,
              target: ctTarget,
              content: ctContent,
            }),
          );
          NavigationActions.showSnackbar(
            `Added capture template with name ${text}`,
          );
        } else {
          const { action } = yield call(
            DialogAndroid.alert,
            'Add Capture Template',
            `Capture template named "${text}" already exists. Do you want to overwrite it ?`,
            {
              positiveText: 'Yes',
              negativeText: 'No',
            },
          );
          if (action === DialogAndroid.actionPositive) {
            yield put(
              CaptureActions.addCaptureTemplate.success({
                name: text,
                target: ctTarget,
                content: ctContent,
              }),
            );

            NavigationActions.showSnackbar(
              `Added capture template with name ${text}`,
            );
          }
        }
      }
      break;
    default:
  }
  after && after();
}

function* cancelCapture(action) {
  const { navigator } = action.payload;
  const navigationStackHistory = yield select(
    OrgDataSelectors.navigationStackHistory,
  );
  if (navigationStackHistory[0] !== 'capture') {
    return;
  }

  const link =
    navigationStackHistory.length > 1 ? navigationStackHistory[1] : 'notes';

  Keyboard.dismiss();
  yield call(navigator.handleDeepLink, { link });
}

function* confirmCapture(raction) {
  const { navigator } = raction.payload;
  const { target, name, ...node } = yield select(getFormValues('capture'));

  const type = node.id ? 'edit' : 'new';
  const navigationStackHistory = yield select(
    OrgDataSelectors.navigationStackHistory,
  );

  const link =
    navigationStackHistory.length > 1 ? navigationStackHistory[1] : 'notes';

  let addedNode;

  switch (type) {
    case 'edit':
      yield put(OrgDataRedux.updateNode(node, [node.id], link));
      break;

    case 'new':
      addedNode = (yield call(OrgApi.addNodes, [prepareNoteForDb(node)], {
        ...target,
      }))[0];

      // yield put(OrgDataRedux.updateNode(addedNode));
      yield* NodeSagas.updateNode({ node: addedNode, isNew: true });

      // TODO zbatchować akcje, aby uniknę wielokrotych updatów
      const fileNodes = yield select(
        OrgDataSelectors.getFileNodes(addedNode.fileId),
      );

      // Update idików - dzieje się tylko przy dodawaniu a nie przy updacie
      // Inna sprawa to timestampy - one mogą się zmieniac przy updacie więc spadaje do updateNode
      const newIds = R.pipe(
        R.values,
        R.sortBy(R.prop('position')),
        R.map(R.prop('id')),
      )(fileNodes);
      const newTOC = R.pipe(
        R.filter(R.propEq('level', 1)),
        R.map(R.prop('id')),
      )(fileNodes);
      yield put([
        OrgDataRedux.addToLastCaptured(addedNode),
        OrgDataRedux.updateFileIds(addedNode.fileId, newIds, newTOC),
      ]);
  }

  yield call(navigator.handleDeepLink, { link });

  /* ------------- Create success message ------------- */

  let message;

  if (type === 'edit') {
    message = 'Edited note saved succesfully.';
  } else if (type === 'add' && name) {
    message = `Added note to ${name}.`;
  } else {
    message = 'Added note.';
  }

  yield call(navigator.showSnackbar, {
    text: message,
    textColor: Colors.bg, // optional
    backgroundColor: Colors.primary, // optional
    duration: 'short', // default is `short`. Available options: short, long, indefinite
  });
}

export function* saveCaptureTemplate({ target, after }) {
  let ctTarget;
  let ctContent;
  let ctName;
  if (target.target) {
    ctTarget = target.target;
    ctContent = R.omit(['target', 'name'], target);
    ctName = target.name;
  } else {
    // Only target for new template
    ctTarget = target;
    ctContent = {};
  }

  yield put(
    CaptureActions.addCaptureTemplate.success({
      name: ctName,
      target: ctTarget,
      content: ctContent,
    }),
  );
}

// * Exports

export default {
  addCaptureTemplate,
  cancelCapture,
  confirmCapture,
  saveCaptureTemplate,
};
