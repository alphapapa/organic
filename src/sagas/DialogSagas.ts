// * DialogSagas.ts

// ** License

/**
 * Copyright (C) 2018, Bart≈Çomiej Nankiewicz<bartlomiej.nankiewicz@gmail.com>
 *
 * This file is part of Organic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// ** Tasks

// *** TODO Add types

// * Imports

import { OrgApi } from 'org-mode-connection';
import R from 'ramda';
import DialogAndroid from 'react-native-dialogs';
import { getFormValues, initialize } from 'redux-form';
import { call, put, select } from 'redux-saga/effects';
import { AgendaSelectors } from '../redux/AgendaRedux';
import CaptureActions, { CaptureSelectors } from '../redux/CaptureRedux';
import NavigationRedux, { NavigationSelectors } from '../redux/NavigationRedux';
import OrgSearcherRedux from '../redux/OrgSearcherRedux';
import OrgDataRedux from '../redux/OrgDataRedux';
import OrgSearcherFilterRedux, {
  SearchFilterSelectors,
} from '../redux/OrgSearcherFilterRedux';
import { vibrate } from '../vibrations';
import { textInputDialog, confirmDialog } from './Helpers';
import NotebookSagas from './NotebookSagas';
import { CaptureTemplate } from '../redux/types';

// * Code

export function* selectCaptureTemplate(action) {
  const { id } = action.payload;
  let ct;

  // Null is used when adding to note to currently visited node
  if (id !== null) {
    ct = yield select(CaptureSelectors.captureTemplate(id));
    yield put(NavigationRedux.loadBreadcrumbs({ target: ct.target }));
    // yield put(NavigationRedux.selectCaptureTemplate())
  } else {
    ct = yield select(getFormValues('capture'));
  }

  // Agenda Capture
  const cType = yield select(NavigationSelectors.getCaptureType);
  switch (cType) {
    case 'agendaCapture':
      const date = yield select(AgendaSelectors.getSelectedDay);
      ct = R.merge(ct, {
        timestamps: [
          {
            type: 'scheduled',
            date,
          },
        ],
      });
      break;
  }

  yield put(initialize('capture', ct));
}

interface DialogAction {
  payload: {
    item: CaptureTemplate; // item id
    action?: string; // actionId
  };
}

export function* showCaptureTemplateActions(raction: DialogAction) {
  const { item, action } = raction.payload;
  vibrate();
  let actionId;

  if (action) {
    actionId = action;
  } else {
    const { selectedItem } = yield call(
      DialogAndroid.showPicker,
      `Select Action for "${item.name}":`,
      null,
      {
        positiveText: '',
        items: [
          { label: 'Rename', id: 'rename' },
          { label: 'Delete', id: 'delete' },
        ],
      },
    );

    if (selectedItem) {
      actionId = selectedItem.id;
    }
  }

  switch (actionId) {
    case 'rename':
      const userInput = yield* textInputDialog(
        'Rename capture template',
        `Enter new name for capture template "${item.name}"`,
      );
      if (userInput) {
        const exists = yield select(
          CaptureSelectors.captureTemplateExists(userInput),
        );

        if (!exists) {
          yield put(
            CaptureActions.addCaptureTemplate.success({
              name: userInput,
              target: item.target,
              content: R.omit(['target', 'name'], item),
            }),
          );
          yield put(CaptureActions.deleteCaptureTemplate({ name: item.name }));
        } else {
          const { action, text } = yield call(
            DialogAndroid.alert,
            'Rename capture template',
            `Capture template named "${userInput}" already exists. Do you want to overwrite it ?`,
            {
              positiveText: 'Yes',
              negativeText: 'No',
            },
          );
          if (action === DialogAndroid.actionPositive) {
            yield put(
              CaptureActions.addCaptureTemplate.success({
                name: userInput,
                target: item.target,
                content: R.omit(['target', 'name'], item),
              }),
            );
            yield put(CaptureActions.deleteCaptureTemplate(item.name));
          }
        }
      }
      break;

    case 'delete':
      let { action } = yield call(
        DialogAndroid.alert,
        'Delete',
        `Do you really want to delete "${item.name}" capture template `,
        {
          positiveText: 'yes',
          negativeText: 'no',
        },
      );

      switch (action) {
        case DialogAndroid.actionPositive:
          yield put(CaptureActions.deleteCaptureTemplate({ name: item.name }));
          break;
        default:
      }
      break;
    default:
  }
}

export function* showFileActionsDialog({ item }) {
  vibrate();
  let actions = [
    { label: 'Rename', id: 'rename' },
    { label: 'Delete', id: 'delete' },
  ];

  if (item.path) {
    actions = actions.concat([
      { label: 'Force load', id: 'forceLoadFile' },
      { label: 'Force write', id: 'forceWrite' },
    ]);
  } else {
    actions = actions.concat([{ label: 'Export', id: 'export' }]);
  }

  const { action, selectedItem } = yield call(
    DialogAndroid.showPicker,
    `Select Action:`,
    null,
    {
      positiveText: '',
      items: actions,
    },
  );

  if (selectedItem) {
    let res;
    switch (selectedItem.id) {
      case 'forceLoadFile':
        res = yield* confirmDialog(
          'Force load',
          'Local changes will be forgotten. Do you really want to reload file?',
        );
        if (res) {
          yield* NotebookSagas.deleteFile(item);
          yield* NotebookSagas.addOrgFile({ filepath: item.path });
        }
        break;

      case 'export':
        DialogAndroid.alert(
          'Export functionality in not available in this version of app.',
        );
        break;

      case 'rename':
        const userInput = yield* textInputDialog(
          'Rename Notebook',
          `Enter new name for notebook "${item.metadata.TITLE}"`,
        );
        if (userInput) {
          let renamedItem = R.assocPath(['metadata', 'TITLE'], userInput, item);
          renamedItem.isChanged = true;
          yield put(OrgDataRedux.updateFile(renamedItem));
          yield call(
            OrgApi.updateFile,
            renamedItem.id,
            R.omit(['id', 'toc', 'nodesData'], renamedItem),
          );
        }
        break;

      case 'delete':
        res = yield* confirmDialog(
          'Delete',
          'Do you really want to delete file?',
        );
        if (res) {
          yield* NotebookSagas.deleteFile(item);
        }

        break;
      default:
    }
  }
}

export function* showSavedSearchesActions(
  action: ReturnType<typeof OrgSearcherRedux.showSavedSearchesActions>,
) {
  vibrate();
  const { item } = action.payload;

  const { selectedItem } = yield call(
    DialogAndroid.showPicker,
    `Select Action for "${item.name}":`,
    null,
    {
      positiveText: '',
      items: [
        { label: 'Rename', id: 'rename' },
        { label: 'Delete', id: 'delete' },
      ],
    },
  );

  if (selectedItem) {
    switch (selectedItem.id) {
      case 'rename': {
        const userInput = yield* textInputDialog(
          'Rename search',
          `Enter new name for search "${item.name}"`,
        );
        if (userInput) {
          const exists = yield select(
            SearchFilterSelectors.itemExists(userInput),
          );

          if (!exists) {
            yield put(
              OrgSearcherFilterRedux.renameQuery({
                oldName: item.name,
                newName: userInput,
              }),
            );
          } else {
            const { action, text } = yield call(
              DialogAndroid.alert,
              'Rename capture template',
              `Capture template named "${userInput}" already exists. Do you want to overwrite it ?`,
              {
                positiveText: 'Yes',
                negativeText: 'No',
              },
            );
            if (action === DialogAndroid.actionPositive) {
              yield put(
                OrgSearcherFilterRedux.deleteQuery({ name: userInput }),
              );
              yield put(
                OrgSearcherFilterRedux.renameQuery({
                  oldName: item.name,
                  newName: userInput,
                }),
              );
            }
          }
        }
        break;
      }

      case 'delete': {
        const { action } = yield call(
          DialogAndroid.alert,
          'Delete',
          `Do you really want to delete "${item.name}" query`,
          {
            positiveText: 'yes',
            negativeText: 'no',
          },
        );

        switch (action) {
          case DialogAndroid.actionPositive:
            yield put(OrgSearcherFilterRedux.deleteQuery({ name: item.name }));
            break;
        }
        break;
      }
    }
  }
}

// * Export

export default {
  showFileActionsDialog,
  showSavedSearchesActions,
  showCaptureTemplateActions,
  selectCaptureTemplate,
};
