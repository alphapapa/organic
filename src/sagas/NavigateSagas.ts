// * NavigateSagas.ts

// ** License

/**
 * Copyright (C) 2018, Bart≈Çomiej Nankiewicz<bartlomiej.nankiewicz@gmail.com>
 *
 * This file is part of Organic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// ** Tasks

// *** TODO Add types

// * Imports

import { OrgApi } from 'org-mode-connection';
import { call, put } from 'redux-saga/effects';
import CaptureActions from '../redux/CaptureRedux';
import LoadSagas from './LoadSagas';

// * Code

export function* openCaptureScreen({ navigator, fileId, nodeId }) {
  const data = {
    id: 'node is',
    headline: 'added node',
    content: '',
    tags: [],
    level: 1,
    position: 0,
  };
  yield put(CaptureActions.captureSuccess(data));
}

export function* openOrgFile({ fileId, nodeId }) {
  if (fileId) {
    const { fileData, nodesList } = yield call(
      OrgApi.getFileAsPlainObject,
      fileId,
    );
    yield* LoadSagas.loadNodes(nodesList);
  }
}

function* visitCaptureTemplate(action) {
  const { item: { target }, navigator } = action.payload;
  let ctNodeId;
  if (target.headline) {
    const nodeId = yield* LoadSagas.getOrCreateNode(target);
    ctNodeId = nodeId;
  }

  const props = {
    screen: 'OrgFileBrowserScreen',
    passProps: {
      contentFoldingLevel: 1,
      fileId: target.fileId,
      foldingLevel: 1,
      navigationStack: 'notes',
      nodeId: ctNodeId,
    },
  };

  navigator.push(props);
}

// * Export

export default {
  CaptureActions,
  openOrgFile,
  visitCaptureTemplate,
};
