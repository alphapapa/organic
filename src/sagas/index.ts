// * index.ts<sagas>

// ** License

/**
 * Copyright (C) 2018, Bart≈Çomiej Nankiewicz<bartlomiej.nankiewicz@gmail.com>
 *
 * This file is part of Organic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// ** Tasks

// *** TODO Add types

// *** TODO Refactor OrgData actions to typesafe actions

// * Imports

import { all, takeEvery, takeLatest } from 'redux-saga/effects';
import CaptureRedux from '../redux/CaptureRedux';
import NavigationRedux from '../redux/NavigationRedux';
import { OrgDataTypes } from '../redux/OrgDataRedux';
import SearchFilterRedux from '../redux/OrgSearcherFilterRedux';
import OrgSearcherRedux from '../redux/OrgSearcherRedux';
import StartupRedux from '../redux/StartupRedux';
import SyncRedux from '../redux/SyncRedux';
import CaptureSagas from './CaptureSagas';
import DialogSagas from './DialogSagas';
import LoadSagas from './LoadSagas';
import NavigateSagas from './NavigateSagas';
import NotebookSagas from './NotebookSagas';
import SearchSagas from './SearchSagas';
import SyncSagas from './SyncSagas';
import SystemSagas from './SystemSagas';
import NodeSagas from './NodeSagas';
import { getType } from 'typesafe-actions';

// * Code

export default function* root() {
  yield all([
    // some sagas only receive an action
    takeLatest(getType(StartupRedux.startup), SystemSagas.startup),

    /* ------------- data ------------- */
    takeLatest(OrgDataTypes.GET_ORG_DATA_REQUEST, LoadSagas.loadOrgData),
    takeLatest(OrgDataTypes.ADD_ORG_FILE_REQUEST, NotebookSagas.addOrgFile),
    takeLatest(OrgDataTypes.ADD_NOTEBOOK_REQUEST, NotebookSagas.addNotebook),
    takeLatest(OrgDataTypes.CLEAR_DB_REQUEST, SyncSagas.clearDb),
    takeLatest(OrgDataTypes.SYNC_ALL_FILES_REQUEST, SyncSagas.syncAllFiles),
    takeLatest(
      getType(SyncRedux.updateExternallyChangedFiles.request),
      SyncSagas.getExternallyChangedFiles,
    ),
    takeLatest(
      OrgDataTypes.SHOW_FILE_ACTIONS_DIALOG,
      DialogSagas.showFileActionsDialog,
    ),

    /* ------------- browser ------------- */
    takeLatest(
      OrgDataTypes.NAVIGATION_STACK_CHANGED,
      SystemSagas.manageBackHandlerFIXremoveBackHandlers,
    ),

    takeLatest(
      getType(NavigationRedux.loadBreadcrumbs),
      LoadSagas.loadBreadcrumbs,
    ),
    takeLatest(OrgDataTypes.VISIT_PLACE, SystemSagas.manageBackHandlerFIX),
    takeLatest(OrgDataTypes.OPEN_FILE_REQUEST, NavigateSagas.openOrgFile),
    takeLatest(OrgDataTypes.LOAD_RELATED_NODES, LoadSagas.loadRelatedNodes),
    takeLatest(OrgDataTypes.UPDATE_NODE, NodeSagas.updateNode),
    takeLatest(OrgDataTypes.LOAD_TOC, LoadSagas.loadTOC),
    takeLatest(OrgDataTypes.RUN_NODE_ACTION_REQUEST, NodeSagas.runNodeAction),
    takeLatest(OrgDataTypes.TOGGLE_MARK, NodeSagas.toggleMark),
    takeLatest(OrgDataTypes.LOAD_WEEK_AGENDA, LoadSagas.loadAgenda),
    takeLatest(OrgDataTypes.LOAD_DAY_AGENDA, LoadSagas.loadAgenda),

    /* ------------- capture ------------- */
    takeLatest(
      getType(CaptureRedux.visitCaptureTemplate),
      NavigateSagas.visitCaptureTemplate,
    ),
    takeLatest(
      getType(NavigationRedux.selectCaptureTemplateRequest),
      DialogSagas.selectCaptureTemplate,
    ),
    takeLatest(getType(CaptureRedux.confirm), CaptureSagas.confirmCapture),
    takeLatest(
      getType(CaptureRedux.addCaptureTemplate.request),
      CaptureSagas.addCaptureTemplate,
    ),
    takeEvery(getType(CaptureRedux.cancelCapture), CaptureSagas.cancelCapture),
    takeLatest(
      getType(CaptureRedux.showCaptureTemplateActions),
      DialogSagas.showCaptureTemplateActions,
    ),

    /* ------------- search ------------- */
    takeLatest(
      getType(OrgSearcherRedux.showSavedSearchesActions),
      DialogSagas.showSavedSearchesActions,
    ),
    takeLatest(getType(OrgSearcherRedux.search), SearchSagas.search),
    takeLatest(
      getType(SearchFilterRedux.saveSearchQuery.request),
      SearchSagas.saveSearch,
    ),
  ]);
}
