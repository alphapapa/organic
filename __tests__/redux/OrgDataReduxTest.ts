import sagas from "../../src/redux/OrgDataRedux";

describe("OrgDataRedux", () => {
  it("should match a snapshot", () => {
    const methods = Object.getOwnPropertyNames(sagas)
    expect(methods).toMatchSnapshot()
  });
});
