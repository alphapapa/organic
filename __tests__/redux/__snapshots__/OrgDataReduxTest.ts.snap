// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrgDataRedux should match a snapshot 1`] = `
Array [
  "addNotebook",
  "addNotebookRequest",
  "addOrgFileRequest",
  "addOrgFileSuccess",
  "addToLastCaptured",
  "clearDbRequest",
  "clearDbSuccess",
  "clearLastAddedNode",
  "deleteFile",
  "deleteNodes",
  "getOrgDataRequest",
  "getOrgDataSuccess",
  "loadDayAgenda",
  "loadRelatedNodes",
  "loadToc",
  "loadWeekAgenda",
  "mergeNodeChanges",
  "move",
  "navigationStackChanged",
  "openFileRequest",
  "openIds",
  "removeNodes",
  "rescheduleAgenda",
  "reset",
  "resetStack",
  "runNodeActionRequest",
  "runNodeAction",
  "setMode",
  "showFileActionsDialog",
  "stackLoaded",
  "syncAllFilesRequest",
  "syncAllFilesSuccess",
  "toggleMark",
  "updateFile",
  "updateFileIds",
  "updateLoadedNodes",
  "updateNode",
  "updateOrgData",
  "updateTimestamps",
  "visitPlace",
]
`;
